Added blog post about using CodePush on Windows (#437)
microsoft/react-native-windows-samples: website/blog/2021-05-17-codepush.md

---

CodePush is a service built by Microsoft, now incorporated into the [Visual Studio App Center platform](https://appcenter.ms/), which provides a way for React Native and Cordova developers to be more agile in delivering updates to their applications.
Thanks to CodePush, you are able to deliver a new bundle (which might contain updated JavaScript files, new images, etc.) without having to fully redeploy a new package. This means, for example, that you can update the content of your application published on a Store without submitting a new package and waiting for the certification process to be completed; or that you can add a new feature in your LOB app without having to redeploy it via Intune or another MDM solution to all your employees.
The only caveat is that **the feature works only for updates which involves the bundle**, which means changes in the JavaScript code, a new asset, etc. If the update involves changes in the binary package (like adding a new capability in the manifest or integrating a new native module), then you are required to submit a new package using the standard approach. 

Enabling Code Push requires to interact with three different layers:

> 1. App Center, which is the cloud platform that will host and deliver your updates. You're going to push new bundles to App Center using the [App Center CLI](https://docs.microsoft.com/en-us/appcenter/cli/).

Reviewer: ```suggestion
1. App Center, which is the cloud platform that will host and deliver your updates. You're going to push new bundles to App Center using the [App Center CLI](https://docs.microsoft.com/appcenter/cli/).
```

## Setting up App Center

The first task is to create a new application in Visual Studio App Center. Thes starting point is the [App Center website](https://appcenter.ms/), where you can login to your dashboard. If you don't already have an account, you can create a free one the first time you login using your Microsoft Account, work account, GitHub, Facebook or Google account.
Once you are logged in, click on **Add new** and choose **Add new app**.

This is the first step where we have to apply a workaround. In this panel, you will be able to setup your application, by specifing the name and the target platform:

![](assets/2021-05-17-codepush/NewApp.png)

> However, as previously mentioned, if you choose **Windows** you won't see React Native as an option. As such, we have to trick App Center to think that our Windows application is actually an Android one.

Reviewer: Is this is a point-in-time thing / do we have a line of sight to fixing this?

Author: It's actually both. Yes, the post is meant to be a "point-in-time" snapshot to unlock who wants to adopt CodePush today. At the same time, App Center is currently in "slow down mode", so I'm not sure if and when Windows support for React Native might be picked up.

**Important!** Before generating a release package, open the manifest of the app (double click on the Package.appxmanifest file in Visual Studio), move to the **Packaging** section and change the **Package name** field to a more friendly name.
By default, the package name assigned to a new UWP app is a GUID, which makes the **Package Family Name** (the unique identifier of a packaged app) very long. Since the Package Family Name is used also to generate the folder where Windows stores the local data (including the updates downloaded from App Center), a very long name could lead CodePush to break and you might get errors like `The filename or extension is too long` when the updated bundle is being installed.

## Integrating the JavaScript APIs

Now we are ready to start setting up CodePush in the JavaScript layer of the application. The key to enable it is to wrap the main `App` class or function (based on the approach you're using) inside the `codePush` one, like in the following examples:

```javascript
//class component
import codePush from "react-native-code-push";

class App extends Component {

>     //your app component

Reviewer: ```suggestion
    // your app component
```

                break;
            case codePush.SyncStatus.DOWNLOADING_PACKAGE:
                console.log("Downloading update.");
                break;
            case codePush.SyncStatus.INSTALLING_UPDATE:
                console.log("Installing update.");
                break;
            case codePush.SyncStatus.UP_TO_DATE:
                console.log("Your app is up-to-date.");
                break;
            case codePush.SyncStatus.UPDATE_INSTALLED:
                console.log("Update installed.");
                break;
        }
    }

    codePushDownloadDidProgress(progress) {
        console.log(progress.receivedBytes + " of " + progress.totalBytes + " received.");
    }
}
App = codePush(MyApp);
export default App;
```

The `codePushStatusDidChange()` event is raised when the status of the update changes. It can assume one of the values exposed by the `SyncStatus` enumerator. The `codePushDownloadDidProgress()` event, instead, is raised when the download has started and, through the `progress` object, you can get the current status, in case you want to display a progress bar or another similar control.

## Wrapping up

> Thanks to CodePush, you can be more agile in keeping your application up-to-date and deliver new features without having to redeploy a completely new package. Currently, App Center doesn't fully support the React Native for Windows implementation, but with a few workarounds you can still leverage it. I highly invite you to read [the whole README of the main repository](https://github.com/microsoft/react-native-code-push/blob/master/README.md), since it contains many additional information for advanced scenarios, like how to handle Store policies, how to manage multiple deployment assignments, etc.

Reviewer: ```suggestion
Thanks to CodePush, you can be more agile in keeping your application up-to-date and deliver new features without having to redeploy a completely new package. Currently, a few workarounds are needed for App Center to integrate with the React Native for Windows implementation. I highly recommend you to read [the whole README of the main repository](https://github.com/microsoft/react-native-code-push/blob/master/README.md), since it contains a lot of additional information for advanced scenarios, like how to handle Store policies, how to manage multiple deployment assignments, etc.
```
